### PR TITLE
passwordstore: Add configurable locking

### PR DESCRIPTION
**Pull request by grembo**
##### SUMMARY
Passwordstore cannot be accessed safely in parallel, which causes
various issues:

- When accessing the same path, multiple different secrets are
  returned when the secret didn't exist (missing=create).
- When accessing the same _or different_ paths, multiple pinentry
  dialogs will be spawned by gpg-agent sequentially, having to enter
  the password for the same gpg key multiple times in a row.
- Due to issues in gpg dependencies, accessing gpg-agent in parallel
  is not reliable, causing plays to fail (this can be fixed by adding
  `auto-expand-secmem` to _~/.gnupg/gpg-agent.conf_ though).

These problems have been described in various github issues in the past,
e.g., https://github.com/ansible/ansible/issues/23816 and https://github.com/ansible/ansible/issues/27277.

This cannot be worked around in playbooks by users in a non-error-prone
way.

It is addressed by adding new configuration options:

- lock:
  - readwrite: Lock all operations
  - write: Only lock write operations (default)
  - none: Disable locking
- locktimeout: Time to wait for getting a lock (s/m/h suffix)
  (defaults to 15m)

These options can also be set in ansible.cfg, e.g.:

    [passwordstore_lookup]
    lock=readwrite
    locktimeout=30s

Also, add a note about modifying gpg-agent.conf.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
passwordstore
plugins/lookup/passwordstore

##### ADDITIONAL INFORMATION
See also #4192 for a discussion that lead to this PR.

### CODE REVIEW

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [inline plugins/lookup/passwordstore.py:149]:
```suggestion
ansible.cfg: |
```
`>` removes line breaks. That will result in a single long line, which is clearly not what you want.

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [inline plugins/lookup/passwordstore.py:92]:
```suggestion
          - none
        version_added: 4.5.0
```

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [inline plugins/lookup/passwordstore.py:99]:
```suggestion
        default: 900
        version_added: 4.5.0
```

Comment (felixfontein) [inline plugins/lookup/passwordstore.py:99]: Is 900 seconds (15 minutes) a good choice? It seems excessively long to me.

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [inline plugins/lookup/passwordstore.py:93]: I think it's easier to read and write with an underscore:
```suggestion
      lock_timeout:
```

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [inline plugins/lookup/passwordstore.py:94]:
```suggestion
        description:
          - Lock timeout in seconds.
          - Used when I(lock) is not C(none).
```

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [inline plugins/lookup/passwordstore.py:83]:
```suggestion
        description:
          - How to synchronize operations.
          - ...you should describe what the values do here...
```

File: plugins/lookup/passwordstore.py
Comment (grembo) [reply plugins/lookup/passwordstore.py:93]: Nothing else in that module uses underscores, hence the choice, examples:
- passwordstore (instead of password_store)
- returnall (instead of return_all)
- subkey (debatable)
- userpass (user_pass?)
- nosymbols (no_symbols?)

Otherwise, I would have preferred lock_timeout as well.

File: plugins/lookup/passwordstore.py
Comment (grembo) [reply plugins/lookup/passwordstore.py:99]: That is based on personal experience:
- Run playbook
- Switch desktops
- A few minutes later switch back, realize that pinentry is waiting for a passphrase to be entered, as gpg-agent removed the key from the cache

Happened frequently enough to decide to use a pretty long timeout.

File: plugins/lookup/passwordstore.py
Comment (grembo) [reply plugins/lookup/passwordstore.py:149]: Thanks. That was just copy and pasted from [here](https://github.com/ansible-collections/community.general/blob/00cab64f7abb687cc2abf600843477e3f4484308/plugins/callback/diy.py#L605) without thinking.

File: plugins/lookup/passwordstore.py
Comment (grembo) [reply plugins/lookup/passwordstore.py:94]: Cool

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [reply plugins/lookup/passwordstore.py:106]: Makes sense. And since this is configurable, if someone dislikes the long timeout (or thinks it is too short :D ) can change it.

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [reply plugins/lookup/passwordstore.py:98]: Fair point.

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [inline plugins/lookup/passwordstore.py:18]:
```suggestion
      - To avoid problems when accesing multiple secrets at once, add C(auto-expand-secmem) to
        C(~/.gnupg/gpg-agent.conf). Where this isn't possible, consider using C(lock=readwrite) instead.
```

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [inline plugins/lookup/passwordstore.py:87]:
```suggestion
          - The default of C(write) only synchronizes write operations.
          - C(readwrite) synchronizes all operations (including read). This makes sure that gpg-agent is never called in parallel.
          - C(none) does not do any synchronization.
```

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [inline plugins/lookup/passwordstore.py:100]:
```suggestion
          - Lock timeout in seconds.
          - The default is 15 minutes.
```

**Comment by grembo** [comment]
@felixfontein Failed test due to some spurious build pipeline issue:

```
07:44 Stopped running freebsd/12.0 instance ************************************.
07:44 ERROR: Command "ssh -tt -q -i /home/AzDevOps_azpcontainer/.ansible/test/id_rsa -o BatchMode=yes -o ServerAliveCountMax=4 -o ServerAliveInterval=15 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -p 22 ec2-user@ec2-34-231-110-215.compute-1.amazonaws.com su -l root -c ''"'"'chmod +x /tmp/remote.sh && /tmp/remote.sh freebsd 3.6'"'"''" returned exit status 56.
##[error]Bash exited with code '1'.
```

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [reply plugins/lookup/passwordstore.py:100]: Specifying what 900 seconds means in minutes is mostly for convenience, since there's a significant amount of people who cannot do this in their head. I don't insist on this though, it's just a suggestion :)

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [inline plugins/lookup/passwordstore.py:392]:
```suggestion
            'lock': self.get_option('lock'),
            'locktimeout': self.get_option('locktimeout'),
```

File: changelogs/fragments/4194-configurable-passwordstore-locking.yml
Comment (felixfontein) [inline changelogs/fragments/4194-configurable-passwordstore-locking.yml:2]:
```suggestion
  - passwordstore lookup plugin - add configurable ``lock`` and ``locktimeout`` options to avoid race conditions in itself and in the ``pass`` utility it calls. By default, the plugin now locks on write operations (https://github.com/ansible-collections/community.general/pull/4194).
```

File: plugins/lookup/passwordstore.py
Comment (grembo) [reply plugins/lookup/passwordstore.py]: @felixfontein To me converting seconds is second - we couldn't afford parsing strings back in the 90s :)

Changed it now so that it requires a suffix to specify the unit (decided against a default, as this is also a matter of taste and potential confusion when reading configs).

**Comment by felixfontein** [comment]
Yes, the Python 3.7 import tests for ansible-core 2.11 currently fail. Simply ignore them.

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [inline plugins/lookup/passwordstore.py:101]:
```suggestion
          - Time with a unit suffix, C(s), C(m), C(h) for seconds, minutes, and hours, respectively. For example, C(900s) equals C(15m).
```

File: plugins/lookup/passwordstore.py
Comment (grembo) [reply plugins/lookup/passwordstore.py:402]: This is handled in parse_terms line 168 (all input validation happens ins that function):

‘’’
if not re.match('^[0-9]+[smh]$', self.paramvals['locktimeout']):
                 raise AnsibleError("{0} is not a correct value for locktimeout".format(self.paramvals['locktimeout']))
‘’’

File: plugins/lookup/passwordstore.py
Comment (grembo) [reply plugins/lookup/passwordstore.py:101]: I used quotes there in purpose (for readability and to distinguish it from the defined values and examples, as in `unit suffixes like “mm”, “cm”, e.g., C(20mm))`

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [reply plugins/lookup/passwordstore.py:402]: Then why not convert the value there? I think it would be a lot cleaner if there's only one piece of code that handles the parsing.

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [inline plugins/lookup/passwordstore.py:265]: This condition can never be true, since you use `self.get_option` to retrieve the value of `lock`.

Comment (grembo) [reply plugins/lookup/passwordstore.py:265]: @felixfontein couldn’t it get overridden by additional terms in the lookup later (run iterates over terms)?

File: plugins/lookup/passwordstore.py
Comment (grembo) [reply plugins/lookup/passwordstore.py:402]: > Then why not convert the value there? I think it would be a lot cleaner if there's only one piece of code that handles the parsing.

I disagree that it would be *a lot* cleaner. This way, we don’t introduce additional variables to be parsed to, and it’s consumed in exactly the place which knows how to interpret the values.

Please feel free to refactor the code later, I can’t spend much more time on this.

**Comment by grembo** [comment]
@felixfontein Looking at the code again, I cleaned up locking and put it all in one function, so locking configuration isn't part of `self.paramvals` anymore. Much less confusing, as `parse_params` would never alter it based on `terms` anyway. I'm much happier with this iteration of the code.

Do you have a final list of things that should change to make this acceptable to you?

Thanks again for your time spent on reviewing this change.

p.s. I did this in a separate commit this time, so it should be easier to follow. Let me know if I should to squash it before it's merged, or if you'll do that.

File: plugins/lookup/passwordstore.py
Comment (grembo) [reply plugins/lookup/passwordstore.py]: In the spirit of wrapping up this contribution without further ado, I adopted your preferred wording.

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [inline plugins/lookup/passwordstore.py:18]:
```suggestion
        C(~/.gnupg/gpg-agent.conf). Where this is not possible, consider using I(lock=readwrite) instead.
```

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [inline plugins/lookup/passwordstore.py:373]: Doing the conversion inside `opt_lock` is really not a good idea. Such things should really be done on setup time. If the user for example sets `lock=write` and an invalid value for `locktimeout`, the plugin already performed a read operation before finally telling the user that `locktimeout` is bad, or might not tell the user at all because no write was needed.

**Comment by felixfontein** [comment]
> p.s. I did this in a separate commit this time, so it should be easier to follow. Let me know if I should to squash it before it's merged, or if you'll do that.

See https://github.com/ansible-collections/community.general/blob/main/CONTRIBUTING.md#open-pull-requests:

> Do not squash your commits and force-push to your branch if not needed. Reviews of your pull request are much easier with individual commits to comprehend the pull request history. All commits of your pull request branch will be squashed into one commit by GitHub upon merge.

File: plugins/lookup/passwordstore.py
Comment (grembo) [reply plugins/lookup/passwordstore.py:18]: What is the logic of using I instead of C in this case?

File: plugins/lookup/passwordstore.py
Comment (grembo) [reply plugins/lookup/passwordstore.py:373]: I’m aware of that,  but I didn’t consider it to be problem - after all, read operations aren’t destructive and if locking isn’t required, the timeout doesn’t matter either.

File: plugins/lookup/passwordstore.py
Comment (grembo) [reply plugins/lookup/passwordstore.py]: Ok, found the documentation... https://docs.ansible.com/ansible/2.5/dev_guide/developing_modules_documenting.html#formatting-functions

File: plugins/lookup/passwordstore.py
Comment (grembo) [reply plugins/lookup/passwordstore.py:373]: Refactored once again

**Comment by grembo** [comment]
> > p.s. I did this in a separate commit this time, so it should be easier to follow. Let me know if I should to squash it before it's merged, or if you'll do that.
> 
> See https://github.com/ansible-collections/community.general/blob/main/CONTRIBUTING.md#open-pull-requests:
> 
> > Do not squash your commits and force-push to your branch if not needed. Reviews of your pull request are much easier with individual commits to comprehend the pull request history. All commits of your pull request branch will be squashed into one commit by GitHub upon merge.

@felixfontein Ok, so all is good.

File: plugins/lookup/passwordstore.py
Comment (felixfontein) [reply plugins/lookup/passwordstore.py]: A more up-to-date version is https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
I will merge this tomorrow if nobody complains.

**Comment by grembo** [comment]
> I will merge this tomorrow if nobody complains.

Thanks

**Comment by felixfontein** [comment]
@grembo thanks for your contribution!

**Pull request merged**
